### ISSUE TITLE
FFI comparible for emscripten/make APIs public

### ISSUE THREAD

**Issue by FrozenSource** [emscripten-core/emscripten#20440]
Hi All,

**Version of emscripten/emsdk:**
Docker emsdk:
emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 3.1.47 (431685f05c67f0424c11473cc16798b9587bb536)

We have been using your API in a not so nice way I'll admit. But we use some internals to dynamically call JS functions. It is a cross-platform project in which on windows we use libffi and here we use this as a comparison. My problem was/is that val doesn't have it's _em_call en EM_VAL constructors public causing me to resort to these internals in the first place.

Now up until now this has worked great, however as of 3.1.47 you guys did something to val causing it to align to 8 bytes instead of pointer 4 on wasm32. As such the current way does not work. I have tried some more hacky stuff with offsets to fix it but to no affail.

As such I have come here on my knees to ask you guys to please make these APIs public or an abstraction on it to allow this dynamic calling?
Or should you have a better solution available.

Any other explanation I'll be able to provide as the snippet is off course rather small...

![IntelliJ Snippet](https://github.com/emscripten-core/emscripten/assets/14978036/9007305b-1fab-4b2d-b9a8-178aed7874df)

Sincerely,
FrozenSource

**Comment by sbc100** [comment]
@RReverser who knows more about EM_VAL that I do.

I can't see anything about `EM_VAL` that changed in 3.1.47.. `EM_VAL` is just a pointer type right?  So it should always have pointer size and alignment.   `emscripten::val` did change though since we added a thread handle there: #20344 .  I suspect that what is effecting you?

(BTW, when you share code snippets is much better for use if you can do it in the form of text rather than images.  Images cannot be copied or searched).

**Comment by RReverser** [comment]
> So it should always have pointer size and alignment. emscripten::val did change though since we added a thread handle there: https://github.com/emscripten-core/emscripten/pull/20344 . I suspect that what is effecting you?

Yeah looks like they're casting `emscripten::val*` / relying on its layout in the example above?

> We have been using your API in a not so nice way I'll admit. But we use some internals to dynamically call JS functions. It is a cross-platform project in which on windows we use libffi and here we use this as a comparison. My problem was/is that val doesn't have it's _em_call en EM_VAL constructors public causing me to resort to these internals in the first place.

Hmm yeah please don't rely on internals - we are likely to change those a lot, as evidenced above, especially in the nearest time. `emscripten::val` is _the_ public API for this stuff.

**Comment by RReverser** [comment]
> But we use some internals to dynamically call JS functions.

`emscripten::val` already calls JS functions dynamically by itself. Can you maybe show example of what you're trying to do and I could show how to do that without touching internals?

**Comment by RReverser** [comment]
It sounds / seems like you want to call JS functions with dynamic list of arguments? If so, you should use JS [`.apply`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/apply) on the `val` representing the function.

```cpp
// create array for dynamic arguments
val args = val::array();
// populate that array
for (/* ... your input args ... */) {
  args.call<void>("push", yourDynamicArgument());
}
// get the function
val func = /* function you want to call */;
// call `.apply` on that function with empty `val` (`undefined`) and the array of arguments
val result = func.call<val>("apply", val(), args);
```

**Comment by FrozenSource** [comment]
Off course I want to get rid of using your internals. But as opposed to the sample above the _em_call function this one is compiled based on templates.

I work on a production programming language which we try to port to webassembly and we’d love to have this feature (calling JS). But it would require complete dynamic calling like we’ve done up till now.

You see we don’t know what the end programmer’s function definition is.

**Comment by RReverser** [comment]
> But as opposed to the sample above the _em_call function this one is compiled based on templates.

It's not, it can accommodate any dynamic number of args. It doesn't matter how you add them to the array.

**Comment by RReverser** [comment]
Judging from your snippet, you already populate `types` and `args` vectors somewhere before the provided code; just change that place to populate `args` from my example instead.

If you are still running into problems, it might be helpful if you posted the whole function.

**Comment by FrozenSource** [comment]
Let me rephrase the variable types and references would be build up in the template. In this specific example I'd have to build a switch case in that for loop and push them in that way. What I liked so much about the _em_call was that I only needed a type lookup table and memcpy to a em_wtype array. Which makes it a whole lot dynamic.

Any way you'd be willing to make a public version of em_call returning a emscripten::val?

**Comment by RReverser** [comment]
> In this specific example I'd have to build a switch case in that for loop and push them in that way.

No, you don't need a switch, just pass them to `push` directly. `val` takes care of the conversions.

Please try to implement it like in my snippet above and let me know if you run into issues. 

> Any way you'd be willing to make a public version of em_call returning a emscripten::val?

Very unlikely, as I said, those are internal implementation details and really shouldn't be used. I'm working on a refactoring that will likely remove `_emval_call` altogether.

**Comment by FrozenSource** [comment]
You prop missed my code report, but as you can see we don't have access to the native type so to say. Now you might say just refactor 3 different platforms worth of code but I think the public version of emval_call might be a great addition.

I'd like to stay with your latest branches if possible and doing this through public channels.

**Comment by FrozenSource** [comment]
obviously I agree with your comments about the internal:: namespace.
But I see no reason whatsoever to get rid of things which make a framework such as emscripten less dynamic and in some case easy to use.
Templating is nice and all but sometimes you just need a c-like api to do your stuff best.
I guess it's just agree to disagree in this case and we'll see what we'll do.

**Comment by RReverser** [comment]
> But I see no reason whatsoever to get rid of things which make a framework such as emscripten less dynamic and in some case easy to use.

You might not see it, but it's important for performance of the public API and being able to use it in more places.

`val` represents JavaScript values, it can't really get more dynamic than that. That's why I'm recommending using `val` directly together with corresponding JS APIs - your code will be simpler, faster, and yet equally dynamic at the same time. What's not to like (except for having to rewrite a little bit of code once to move away from APIs that weren't meant to be used in the first place).

You even already have correct switch using the `val`'s APIs for the return value - the one for arguments can be pretty much copy-pasted from that, and just invert the direction so that it reads `val`s instead of writing them.

**Comment by RReverser** [comment]
> We have been using your API in a not so nice way I'll admit. But we use some internals to dynamically call JS functions. It is a cross-platform project in which on windows we use libffi and here we use this as a comparison.

Btw you say you're using libffi on other platforms. Just in case - you know there is a port of libffi for Emscripten too? https://github.com/hoodmane/libffi-emscripten

If that works for you, you might not even need to maintain your own implementation and just compile with libffi on all platforms.

**Comment by RReverser** [comment]
Ah looks like it was even upstreamed already, so libffi itself already supports Emscripten: https://github.com/libffi/libffi/tree/master/src/wasm32

**Comment by FrozenSource** [comment]
Hmmmm, that wasn't available yet but we'd lose the string (as far as I can see in ffi.c) type all together which would bring more hassle to bring it back. Thanks for the info either way.

I agree on the greatness of val but I could be entirely mistaken but _em_call should perform better as you'd only need to context switch once from WASM to JS no? (https://github.com/emscripten-core/emscripten/blob/5b63268c904dd0e542f9c45e2b136c3d4a0bd638/src/embind/emval.js#L348)

Either way I just think it is a sad loss of a great function. But I'll just have to buckle then...

**Comment by FrozenSource** [comment]
I'll keep it in mind, thanks!

**Comment by RReverser** [comment]
Great, I'll close this issue for now then, but let me know if you run into problems with `.apply`. As I said, I think it should be fairly straightforward port of your existing `switch` for the return value, but I'm happy to help.
